runners/flink/1.15/src/main/java/org/apache/beam/runners/flink/transform/sql/FlinkSql.java

```diff
@@ -0,0 +1,463 @@
+  private final Map<TupleTag<?>, PValue> additionalInputs;
+  private final List<TupleTag<?>> additionalOutputs;
+  private final Map<String, TableInfo.InputTableInfo<?>> inputTableInfo;
+  private final Map<String, TableInfo.OutputTableInfo<?>> outputTableInfo;
+  private String mainInputTableName;
+  private String mainOutputTableName;
+
+  FlinkSql(Class<InputT> inputClass, Class<OutputT> outputClass) {
+    this.queries = new LinkedHashMap<>();
+    this.ddlList = new ArrayList<>();
+    this.additionalInputs = new HashMap<>();
+    this.additionalOutputs = new ArrayList<>();
+    this.inputTableInfo = new HashMap<>();
+    this.outputTableInfo = new HashMap<>();
+    this.mainInputTableName = UNUSED_MAIN_INPUT_TABLE_NAME;
+    this.mainOutputTableName = DEFAULT_MAIN_OUTPUT_TABLE_NAME;
+    inputTableInfo.put(mainInputTableName, TableInfo.InputTableInfo.of(inputClass));
+    outputTableInfo.put(mainOutputTableName, TableInfo.OutputTableInfo.of(outputClass));
+  }
+
+  /**
+   * Create a {@link SingleOutputSqlTransform} with the specified input and output class. By
+   * default, the created SQL Transform assumes both the input and output classes are of the <a
+   * href="https://en.wikipedia.org/wiki/Plain_old_Java_object">POJO type</a>. So, the SQL transform
+   * can convert a record to a Flink Table Row and vice versa.
+   *
+   * <p>Users can also specify the conversion between a record of Java object and a Flink Table Row
+   * by specifying a {@link TypeInformation} and {@link AbstractDataType DataType} if needed.
+   *
+   * @param inputClass the class of the input records for the Sql transform.
+   * @param outputClass the class of the output records for the sql transform.
+   * @return A {@link SingleOutputSqlTransform}.
+   */
+  public static <InputT, OutputT> SingleOutputSqlTransform<InputT, OutputT> of(
+      Class<InputT> inputClass, Class<OutputT> outputClass) {
+    return new SingleOutputSqlTransform<>(new FlinkSql<>(inputClass, outputClass));
+  }
+
+  // --------------------- setters ----------------------------
+  /**
+   * Use DDL to define Tables. The DDL string can contain multiple {@code CREATE TABLE} / {@code
+   * CREATE VIEW} statements. The DDL string should not contain any DQL / DML statement.
+   *
+   * @param ddl the table definition
+   * @return this {@link FlinkSql} itself.
+   */
+  FlinkSql<InputT, OutputT> withDDL(String ddl) {
+    ddlList.add(ddl);
+    return this;
+  }
+
+  /**
+   * Use DQL to express the query logic. The query should only contain one DQL, i.e. one top level
+   * {@code SELECT} statement. The query result will be registered as a temporary view. The query
+   * statements comes after this query can refer to the result of this query with the specified
+   * result table name.
+   *
+   * @param resultTableName the table name of the query result.
+   * @param query the SQL DQL statement.
+   */
+  FlinkSql<InputT, OutputT> withQuery(String resultTableName, String query) {
```

Comment:
Seems as a Beam API, the resultTableName should be a TupleTag, which contains both the name and type to retrieve the output. I think in most of the use cases, the user will provide a single query, which the resultTableName will be unused. For simplicity, maybe provide a flavor of withQuery(String query) and use the default main output tupleTag?